ros_ws/src/interpreter_pkg/interpreter_pkg/interpreter.py

```diff
@@ -0,0 +1,168 @@
+from rclpy.node import Node
+from example_interfaces.msg import String
+
+import langchain
+from langchain.chains import LLMChain
+from langchain.chat_models import ChatOpenAI
+from langchain.llms import OpenAI
+from langchain.prompts import (ChatPromptTemplate, HumanMessagePromptTemplate,
+                                SystemMessagePromptTemplate)
+from langchain.schema import  HumanMessage, SystemMessage
+
+from ament_index_python.packages import get_package_share_directory
+
+data_dir = os.path.join(get_package_share_directory("interpreter_pkg"))
+OpenAI.api_key=os.environ["OPENAI_API_KEY"]
+MODEL_TO_USE="gpt-4"
+
+class LLM_Object:
+    def __init__(self, context_file: str, model:str, temperature:int):
+        with open(os.path.join(data_dir,context_file), 'r') as file:
+            context= file.read().replace('\n', '')
+        self.context=context
+        print("context is: "+self.context)
+        self.rolling_context=self.context
+        self.temp=temperature
+        self.chat=ChatOpenAI(model_name=model)
+        self.system_message=SystemMessage(content=self.context)
+
+    def reset_context(self):
+        self.rolling_context=[self.system_message]
+        
+class TerpNode(Node):
+    def __init__(self):
+        super().__init__('interpreter')
+        self.get_logger().info("Terp Started")
+        #define elements
+        self.refiner=LLM_Object("refiner_context.prompt",MODEL_TO_USE,0) 
+        self.judger=LLM_Object("judge_context.prompt",MODEL_TO_USE, 0) 
+        self.extractor=LLM_Object("extractor_context.prompt",MODEL_TO_USE,0) 
+        self.new_request=True
+        self.attempts_to_understand=0
+        self.target=""
+        self.rolling_context=[]
+        self.last_suggestion=""
+        # define publishing
+        self.target_publisher_=self.create_publisher(String,'target',10)
+        self.chat_publisher_=self.create_publisher(String, 'clarifying_question',10)
+        #define subscriptions
+        self.subscriber_=self.create_subscription(String, 'user_request',self.get_request,10)
+
+    def get_request(self, request_msg: String): #callback for subscribing to user response
+        #this callback gets the user response to a chat string
+        #check if the response is yes-that means user is confirming your last item,
+        #if so, the last ai message is the target
+        request=request_msg.data
+
+        if request.strip().lower()=="yes" and self.new_request==False:
+            #user is responding to a direct suggestion
+            self.target=extract_request(self.refiner.last_suggestion) 
+            self.broadcast_target(self.target)
+
+        self.new_request=False         
+
+        #SO if not responding yes, use judge to see if the item is actionable in response
+        judgement=judge(self.judger, request)
+        print("data directroy is :"+data_dir)
+        print("judge context is: "+self.judger.context)
+        print("judgement= "+ judgement)
+        if judgement=="yes":
```

Comment:
Do we have issues parsing? IE does the LLM ever produce non yes/no answers?

Reply:
in the judger context, I ask the LLM to only answer yes or no.  At one point in dev, I asked for yes or no and the target, but there was some confusion and I was getting non definitive answers so I created the extractor to extract the target if the judges answer was yes.  With more time I might be able to get them merged into one.  I would think they should be one call.